tests/restapi/test_threaded_restapi.py

```diff
@@ -0,0 +1,87 @@
+# -*- coding: utf-8 -*-
+###########################################################################
+# Copyright (c), The AiiDA team. All rights reserved.                     #
+# This file is part of the AiiDA code.                                    #
+#                                                                         #
+# The code is hosted on GitHub at https://github.com/aiidateam/aiida-core #
+# For further information on the license, see the LICENSE.txt file        #
+# For further information please visit http://www.aiida.net               #
+###########################################################################
+# pylint: disable=protected-access
+"""Tests for the `aiida.restapi` module, using it in threaded mode.
+
+Threaded mode is the default (and only) way to run the AiiDA REST API (see `aiida.restapi.run_api:run_api()`).
+This test file's layout is inspired by https://gist.github.com/prschmid/4643738
+"""
+from threading import Thread
+
+import requests
+from werkzeug.serving import make_server
+
+from aiida.backends.testbase import AiidaTestCase
+from aiida.restapi.api import App, AiidaApi
+from aiida.restapi.common.config import CLI_DEFAULTS, API_CONFIG
+from aiida.restapi.run_api import configure_api
+
+# The URL to the REST API that we will start up. This is the default location
+# where Flask apps start when using the built in web server.
+SERVER_URL = 'http://{hostname}:{port}{api}'.format(
+    hostname=CLI_DEFAULTS['HOST_NAME'], port=CLI_DEFAULTS['PORT'], api=API_CONFIG['PREFIX']
+)
+
+
+class ThreadedRESTApiTestCase(AiidaTestCase):
+    """A TestCase to test the thread safety of the AiiDA REST API"""
+
+    def setUp(self):
+        """Setup to run before each test case."""
+        super(ThreadedRESTApiTestCase, self).setUp()
+
+        # Use the factory method to create the app.
+        _, self.api = configure_api(App, AiidaApi)
+
+        # Create server
+        self.server = make_server(
+            host=CLI_DEFAULTS['HOST_NAME'],
+            port=int(CLI_DEFAULTS['PORT']),
+            app=self.api.app,
```

Comment:
> On a side note (not a problem of your PR, but maybe it can be simplified), since apparently the app is an attribute of the api, why is configure_api returning the tuple and not just the api. If someone needs the app they can get it through api.app. Returning the tuple is a bit pointless.

Good point. Since `configure_api` was introduced very recently (in 1.2.1), it might still be ok to change the return value since I doubt anybody has start using it outside aiida-core.
The return value of `run_api` should not be changed.

Reply:
Agreed, I wondered this myself as well.
And concerning your comment of using `api.app`, again, it is to be as close to the way `verdi restapi` runs the REST API as possible. In the code it runs the `app` exactly this way, i.e., `self.api.app.run()`.

Reply:
Also, I intended to add more tests here, but found I didn't need to, i.e., the `self` is a leftover of this idea.
I will have to redo all this if moving to the `pytest` test design in any case.